rskj-core/src/main/java/co/rsk/RskContext.java

```diff
@@ -909,44 +960,209 @@ protected NodeRunner buildNodeRunner() {
+        checkIfNotClosed();
+
+        if (blockParentDependantValidationRule == null) {
+            Constants commonConstants = getRskSystemProperties().getNetworkConstants();
+            blockParentDependantValidationRule = new BlockParentCompositeRule(
+                    new BlockTxsFieldsValidationRule(),
+                    new BlockTxsValidationRule(getRepositoryLocator()),
+                    new PrevMinGasPriceRule(),
+                    new BlockParentNumberRule(),
+                    new BlockDifficultyRule(getDifficultyCalculator()),
+                    new BlockParentGasLimitRule(commonConstants.getGasLimitBoundDivisor())
+            );
+        }
+
+        return blockParentDependantValidationRule;
+    }
+
+    public synchronized org.ethereum.db.BlockStore buildBlockStore(String databaseDir) {
+        checkIfNotClosed();
+
+        File blockIndexDirectory = new File(databaseDir + "/blocks/");
+        File dbFile = new File(blockIndexDirectory, "index");
+        if (!blockIndexDirectory.exists()) {
+            if (!blockIndexDirectory.mkdirs()) {
+                throw new IllegalArgumentException(String.format(
+                        "Unable to create blocks directory: %s", blockIndexDirectory
+                ));
             }
-            newTrieStore = buildTrieStore(trieStorePath);
         }
-        return newTrieStore;
+
+        DB indexDB = DBMaker.fileDB(dbFile)
+                .make();
+
+        KeyValueDataSource blocksDB = LevelDbDataSource.makeDataSource(Paths.get(databaseDir, "blocks"));
+
+        return new IndexedBlockStore(getBlockFactory(), blocksDB, new MapDBBlocksIndex(indexDB));
+    }
+
+    public synchronized PeerScoringReporterService getPeerScoringReporterService() {
+        checkIfNotClosed();
+
+        if(peerScoringReporterService == null) {
+            this.peerScoringReporterService = PeerScoringReporterService.withScheduler(getRskSystemProperties().getPeerScoringSummaryTime(), getPeerScoringManager());
+        }
+
+        return peerScoringReporterService;
+    }
+
+    public boolean isClosed() {
+        return closed;
+    }
+
+    /**
+     * This method closes this RSK context.
+     *
+     * Internally it stops a node runner, if started,
+     * and closes / disposes data storages (db instances), if some has already been instantiated.
+     *
+     * Note that this method is idempotent, which means that calling this method more than once does not have any
+     * visible side effect.
+     */
+    @Override
+    public synchronized void close() {
+        if (closed) {
+            return;
+        }
+
+        closed = true;
+
+        if (nodeRunner != null) {
+            logger.trace("stopping nodeRunner.");
+            nodeRunner.stop();
+            logger.trace("nodeRunner stopped.");
+        }
+
+        if (blockStore != null) {
+            logger.trace("closing blockStore.");
+            blockStore.close();
```

Comment:
The IndexedBlockStore is not closing the KeyValueDataSource

Reply:
good point. Added a capability to close `IndexedBlockStore`